### PR TITLE
Delete Run UI

### PR DESCRIPTION
**Pull request by andrewmchen**
An initial implementation of the delete run UI. Here are some screenshots,
screen shot 2018-09-19 at 1 55 01 pm
screen shot 2018-09-19 at 1 55 09 pm
screen shot 2018-09-19 at 1 55 17 pm
screen shot 2018-09-19 at 1 55 29 pm

### CODE REVIEW

File: mlflow/server/js/src/Actions.js [hunk 1]
```diff
@@ -52,12 +52,36 @@ export const getRunApi = (runUuid, id = getUUID()) => {
...
 
+export const DELETE_RUN_API = 'DELETE_RUN_API';
+export const deleteRunApi = (runUuid, id = getUUID()) => {
+  return (dispatch) => {
+    const deleteResponse = dispatch({
+      type: DELETE_RUN_API,
+      payload: wrapDeferred(MlflowService.deleteRun, { run_id: runUuid }),
+      meta: { id: getUUID() },
+    });
+    return deleteResponse.then(dispatch(getRunApi(runUuid, id)));
```
Comment (smurching) [inline mlflow/server/js/src/Actions.js]: Ah cool, neat to see this example of how we can chain together promises with redux-promise-middleware & redux-thunk :). I see what you meant earlier about whether `dispatch` will return a promise (i.e. if `deleteResponse` will be a promise) - I don't think `redux-promise-middleware` guarantees this in any API docs, but experimentally it does seem to be the case. [This example](https://github.com/pburtchaell/redux-promise-middleware/blob/master/docs/guides/rejected-promises.md#catching-errors-locally) also suggests that `dispatch` will return a Promise.

Comment (smurching) [reply mlflow/server/js/src/Actions.js]: Do you know what best practices are for error-handling on the `getRunApi` action? [This example](https://github.com/pburtchaell/redux-promise-middleware/blob/master/docs/guides/rejected-promises.md#catching-errors-locally) seems to suggest that we should have error-handling logic in the action creator itself (i.e. not [our existing pattern](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/components/modals/RenameRunModal.js#L64), which relies on e.g. `dispatch(setTagApi())` returning a promise & calls try/catch on that promise in individual components). However I think the right error-handling behavior is often specific to each component, so not sure it makes sense to always handle errors in action creators

File: mlflow/server/js/src/components/ExperimentView.js [hunk 2]
```diff
@@ -59,28 +70,66 @@ class ExperimentView extends Component {
...
+    if (this.state.showRestoreRunModal && nextState.showRestoreRunModal) return false;
+    return true;
+  }
+
+
   static getDerivedStateFromProps(nextProps, prevState) {
-    const { searchInput, paramKeyFilter, metricKeyFilter } = nextProps;
+    // Compute the actual runs selected. (A run cannot be selected if it is not passed in as a
+    // prop)
+    const newRunsSelected = {};
```
Comment (smurching) [inline mlflow/server/js/src/components/ExperimentView.js]: To confirm: is the intent of this to preserve which runs are selected e.g. after a search? It for some reason doesn't seem to work for me

**Review by smurching — COMMENTED** [review]
Thanks @andrewmchen, this is super useful functionality :)

Leaving some comments (plus questions since I'm still a bit of a redux + JS noob :P).

File: mlflow/server/js/src/Actions.js [hunk 3]
```diff
@@ -96,6 +120,21 @@ export const setTagApi = (runUuid, tagName, tagValue, id = getUUID()) => {
   };
 };
 
+export const CLOSE_ERROR_MODAL = 'CLOSE_ERROR_MODAL';
```
[the remark this answers is not in the document]
Comment (andrewmchen) [reply mlflow/server/js/src/Actions.js]: Nice catch, I hadn't seen that before. My preference is to have a follow up PR which converts all modals to use that infra.

File: mlflow/server/js/src/components/ExperimentView.js (hunk 2 above)
Comment (andrewmchen) [reply mlflow/server/js/src/components/ExperimentView.js]: No, it's actually to remove "deleted" runs from the selected set after a deletion happens.

File: mlflow/server/js/src/components/ExperimentView.js [hunk 4]
```diff
@@ -320,14 +421,19 @@ class ExperimentView extends Component {
     this.setState({ searchInput: event.target.value });
   }
 
+  onLifecycleFilterInput(newLifecycleInput) {
+    this.setState({ lifecycleFilterInput: newLifecycleInput });
```
[the remark this answers is not in the document]
Comment (andrewmchen) [reply mlflow/server/js/src/components/ExperimentView.js]: Good catch, I've made the change.

File: mlflow/server/js/src/components/modals/ConfirmModal.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+  onRequestCloseHandler() {
+    if (!this.state.isSubmitting) {
+      this.props.onClose();
+    }
+  }
+
+  handleSubmitWrapper() {
+    this.setState({ isSubmitting: true });
+    this.props.handleSubmit().finally(() => {
```
[the remark this answers is not in the document]
Comment (andrewmchen) [reply mlflow/server/js/src/components/modals/ConfirmModal.js]: Also good call. I've added the babel-polyfill to our build which includes Promise.finally

File: mlflow/server/js/src/components/modals/DeleteRunModal.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  render() {
+    const number = this.props.selectedRunIds.length;
+    return (
+      <ConfirmModal
+        isOpen={this.props.isOpen}
+        onClose={this.props.onClose}
+        handleSubmit={this.handleSubmit}
+        title={`Delete Experiment ${Utils.pluralize("Run", number)}`}
```
[the remark this answers is not in the document]
Comment (andrewmchen) [reply mlflow/server/js/src/components/modals/DeleteRunModal.js]: Just following the designs here! Let's make the change later if we have consensus.

File: mlflow/server/js/src/components/modals/RestoreRunModal.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    dispatch: PropTypes.func.isRequired,
+  };
+
+  handleSubmit() {
+    const restorePromises = [];
+    this.props.selectedRunIds.forEach((runId) => {
+      restorePromises.push(this.props.dispatch(restoreRunApi(runId)));
+    });
+    return Promise.all(restorePromises).catch(() => {
+      this.props.dispatch(openErrorModal('While deleting an experiment run, an error occurred.'));
```
Comment (smurching) [inline mlflow/server/js/src/components/modals/RestoreRunModal.js]: micronit: "While deleting" -> "While restoring"

**Review by smurching — APPROVED** [review]
Awesome, thanks for the updates - LGTM!

**Pull request merged**